Microsoft.Toolkit.Uwp.UI.Controls/Carousel/Carousel.cs

```diff
@@ -0,0 +1,816 @@
+                {
+                    switch (eventArgs.Action)
+                    {
+                        case NotifyCollectionChangedAction.Remove:
+                            foreach (var oldItem in eventArgs.OldItems)
+                            {
+                                for (int i = 0; i < lightStone.internalList.Count; i++)
+                                {
+                                    var fxElement = lightStone.internalList[i] as FrameworkElement;
+                                    if (fxElement == null || fxElement.DataContext != oldItem)
+                                    {
+                                        continue;
+                                    }
+
+                                    lightStone.RemoveAt(i);
+                                }
+                            }
+
+                            break;
+                        case NotifyCollectionChangedAction.Add:
+                            foreach (var newItem in eventArgs.NewItems)
+                            {
+                                lightStone.CreateItem(newItem, 0);
+                            }
+
+                            break;
+                    }
+                };
+            }
+
+            lightStone.Bind();
+        }
+
+        /// <summary>
+        /// Gets or sets item Template
+        /// </summary>
+        public DataTemplate ItemTemplate
+        {
+            get { return itemTemplate; }
+            set { itemTemplate = value; }
+        }
+
+        public Carousel()
+        {
+            this.PointerPressed += OnPointerPressed;
+            this.ManipulationMode = ManipulationModes.TranslateX;
+            this.ManipulationCompleted += OnManipulationCompleted;
+            this.ManipulationDelta += OnManipulationDelta;
+
+            this.Tapped += OnTapped;
+        }
+
+        /// <summary>
+        /// Bind all Items
+        /// </summary>
+        private void Bind()
+        {
+            if (ItemsSource == null || (ItemsSource as IEnumerable) == null)
+            {
+                return;
+            }
+
+            this.Children.Clear();
+            this.internalList.Clear();
+
+            foreach (object item in (IEnumerable)ItemsSource)
+            {
+                this.CreateItem(item);
+            }
+        }
+
+        /// <summary>
+        /// Remove an item at position. Reaffect SelectedIndex
+        /// </summary>
+        public void RemoveAt(int index)
+        {
+            var uiElement = this.internalList[index];
+
+            // this.isUpdatingPosition = true;
+            // Remove from internal list
+            this.internalList.RemoveAt(index);
+
+            // Remove from visual tree
+            this.Children.Remove(uiElement);
+
+            if (this.SelectedIndex == index)
+            {
+                this.SelectedIndex = this.SelectedIndex == 0 ? 0 : this.SelectedIndex - 1;
```

Comment:
If the last item is removed, SelectedIndex should become -1. Here it would end up with 0.